### ISSUE TITLE
Tarsnap could accept keyfiles missing final \n

### ISSUE THREAD

**Issue by np** [Tarsnap/tarsnap#258]
It happened to me that my key files got stripped out of their final new line character. Tarsnap was then considering them as "corrupt or truncated" which is fair.

I was wondering if this would do any harm to be a bit more lenient here.

**Comment by cemeyer** [comment]
Me too, although I disagree with the misleading error message being reasonable.

https://github.com/Tarsnap/tarsnap/blob/master/lib/keyfile/keyfile.c#L300

**Comment by gperciva** [comment]
@cemeyer: Doesn't "truncated" accurately describe a file that's missing the final character?  I suppose you could call it "slightly truncated" instead, but either way the file is shorter than expected.

I'd consider writing a check to warn specifically about a missing final `\n`, but what's the use case?  I mean, how does one lose the final `\n` in a benign manner?  Does it occur via minor disk corruption (which fortunately doesn't disrupt anything else in the keyfile), or by manually editing the keyfile in vim with `:set noeol`, or...?

**Comment by cemeyer** [comment]
@gperciva The "truncation" characterization is only accurate if you believe that the final piece of whitespace, after the ASCII trailer, is a quintessential piece of cryptographic key file material.  And "corrupt" is inaccurate.

Regardless, the error message is unhelpful as-is.  The print could simply add the text "missing newline at end of file."  Or just handle the missing newline gracefully without error.

> I'd consider writing a check to warn specifically about a missing final \n, but what's the use case? I mean, how does one lose the final \n in a benign manner?

Copy (many text applications do not consider trailing newline significant) and paste (via e.g. `xclip -selection clipboard -out > tarsnap.key`).

**Comment by gperciva** [comment]
Hmm.  I see what you mean about copying from a graphical application.

For clarity, `xclip` works by itself:
```
$ xclip ~/.test-tarsnap/tarsnap.keys 
$ xclip -o > test.keys 
$ diff ~/.test-tarsnap/tarsnap.keys test.keys
$ 
```

but this does not:
```
$ gedit ~/.test-tarsnap/tarsnap.keys
       (ctrl-a  ctrl-c, switch to alternate terminal window)
$ xclip -o > test.keys 
$ diff ~/.test-tarsnap/tarsnap.keys test.keys
69c69
< # END OF TARSNAP KEY FILE
---
> # END OF TARSNAP KEY FILE
\ No newline at end of file
$ 
```

(**WARNING** for anybody reading this: don't try this with your real keys unless you're certain that your clipboard cannot be read by any untrusted software.)

As for being accurete in "an extremely literal sense"... yes, a comment line in the Tarsnap key file format must end with EOL.  That's part of the 7-line specification:
https://github.com/Tarsnap/tarsnap/blob/master/lib/keyfile/keyfile.c#L31
Yes, technically @cperciva could change the specs to make an EOL optional, but I'm comfortable going out on a limb and saying that there's no way he would do that.  He's committed to having simple and unambiguous specifications, because he's seen how many security flaws arise from the alternative.

(That said, note that the keyfile doesn't need those comment lines.  They're clearly optional in the definition, and I just double-checked by deleting them and running tarsnap without any problems.)

I think I can convince him to accept a patch that adds a warning specifically for a final missing newline, since that wouldn't touch the cryptographic stuff.  I'll make one and see what he says.

**Comment by cemeyer** [comment]
Right.

A loose BNF-ish in a comment in the source code isn't exactly a user-facing specification.  I.e., nothing in the published documentation could help me determine why tarsnap was producing the error message.  I had to dig into the source because (1) the documentation is lacking and (2) the error messages are devoid of information.  I really don't see how permitting no trailing newline makes a base64-encoded key spec ambiguous.

You could improve any or all of these things without sacrificing any security property.

More broadly than just the EOL/EOF warning, you could add context to all of the myriad identical "corrupt or truncated file" error messages.

None of this affects the security model or functionality of the application; just the UX.

**Comment by gperciva** [comment]
I've sketched two PRs: warn specifically about a missing newline, or accept it.  @cperciva can think about it.

**Comment by cemeyer** [comment]
Fixed in #334, thanks!
